src/libstd/sync/remutex.rs

```diff
@@ -0,0 +1,248 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use prelude::v1::*;
+
+use cell::UnsafeCell;
+use marker;
+use ops::{Deref, DerefMut};
+use sync::poison::{self, TryLockError, TryLockResult, LockResult};
+use sys::mutex as sys;
+use fmt;
+
+/// A re-entrant mutual exclusion
+///
+/// This mutex will block *other* threads waiting for the lock to become available. The thread
+/// which has already locked the mutex can lock it multiple times without blocking, preventing a
+/// common source of deadlocks.
+///
+/// # Examples
+///
+/// ```
+/// use std::sync::ReentrantMutex;
+///
+/// fn locking_mod(data: ReentrantMutex<i32>) {
+///     let mut data = data.lock().unwrap();
+///     *data += 1;
+/// }
+///
+/// let data = ReentrantMutex::new(0);
+/// {
+/// // We take the lock twice here. It won’t deadlock.
+/// let lock = data.lock();
+/// locking_mod(data);
+/// }
+/// ```
+#[unstable(feature = "reentrant_mutex", reason = "new API")]
```

Comment:
You should be able to lift this unstable attribute to the top of the module and then all members will inherit it (e.g. no need to repeat!)